astrocut/ImageCutout.py

```diff
@@ -0,0 +1,428 @@
+            for file, cutout_list in self._cutout_dict.items():
+                if not cutout_list:
+                    warnings.warn(f'Cutout of {file} contains no data and will not be written.', DataWarning)
+                    continue
+                for i, cutout in enumerate(cutout_list):
+
+                    im = Image.fromarray(cutout)
+                    if self._memory_only:
+                        cutout_path.append(im)
+                        continue
+
+                    # Write individual cutouts to disk
+                    file_path = '{}_{:.7f}_{:.7f}_{}-x-{}_astrocut_{}{}'.format(
+                        Path(file).stem,
+                        self._coordinates.ra.value,
+                        self._coordinates.dec.value,
+                        str(self._cutout_size[0]).replace(' ', ''), 
+                        str(self._cutout_size[1]).replace(' ', ''),
+                        i,
+                        self._output_format)
+                    file_path = Path(self._output_dir, file_path).as_posix()
+                    success = self._save_img_to_file(im, file_path)
+                    if success:
+                        cutout_path.append(file_path)
+
+        return cutout_path
+
+    def _write_cutouts(self) -> Union[str, List]:
+        """
+        Write the cutout to a file according to the specified output format.
+
+        Returns
+        -------
+        cutout_path : Path | list
+            Cutouts as memory objects or path(s) to the written cutout files.
+
+        Raises
+        ------
+        InvalidInputError
+            If the output format is not supported.
+        """
+        if self._memory_only:
+            # Write only to memory if specified
+            log.info('Writing cutouts to memory only. No output files will be created.')
+        else:
+            # If writing to disk, ensure that output directory exists
+            Path(self._output_dir).mkdir(parents=True, exist_ok=True)
+
+        if self._output_format == '.fits':
+            return self._write_as_fits()
+        elif self._output_format == '.asdf':
+            return self._write_as_asdf()
+        elif self._output_format in Image.registered_extensions().keys():
+            return self._write_as_img()
+        else:
+            raise InvalidInputError(f'Output format {self._output_format} is not supported.')
+        
+    def cutout(self) -> Union[str, List[str], List[fits.HDUList]]:
+        """
+        Generate cutouts from a list of input images.
+
+        Returns
+        -------
+        cutout_path : Path | list
+            Cutouts as memory objects or path(s) to the written cutout files.
+
+        Raises
+        ------
+        InvalidQueryError
+            If no cutouts contain data.
+        """
+        # Track start time
+        start_time = monotonic()
+
+        # Cutout each input file
+        for file in self._input_files:
+            self._cutout_file(file)
+
```

Comment:
Did astrocut ever use multiprocessing for handling batch cutouts?  Given some of the roman reqs and the tests you did before on mass cutouts, we might want to create optional modes of batch processing files.   Not something for this PR, but maybe something for later?